kernel/core/os.adb

```diff
@@ -933,44 +928,41 @@ is
      (status    : out os_status_t;
       mbx_entry : out os_mbx_entry_t)
    is
-      current        : os_task_id_param_t;
+      --  retrieve current task id
+      current        : constant os_task_id_param_t := os_sched_get_current_task_id;
       mbx_index      : os_mbx_index_t;
       next_mbx_index : os_mbx_index_t;
    begin
       mbx_entry.sender_id := OS_TASK_ID_NONE;
       mbx_entry.msg       := 0;
 
-      --  retrieve current task id
-      current := os_sched_get_current_task_id;
-
       if os_mbx_is_empty (current) then
          --  mbx queue is empty, so we return with error
          status := OS_ERROR_FIFO_EMPTY;
       else
          --  initialize status to error in case we don't find a mbx.
          status := OS_ERROR_RECEIVE;
 
-         --  go through received mbx for this task
-         for iterator in 0 .. (os_mbx_get_mbx_count (current) - 1) loop
+         --  Compute the first mbx_index for the loop
+         mbx_index := os_mbx_get_mbx_head (current);
 
-            --  Compute the mbx_index for the loop
-            mbx_index := os_mbx_get_mbx_head (current) + iterator;
+         --  go through received mbx for this task
+         for iterator in 1 .. os_mbx_get_mbx_count (current) loop
 
             --  look into the mbx queue for a mbx that is waited for
             if os_mbx_is_waiting_mbx_entry (current, mbx_index) then
 
                --  copy the mbx into the task mbx entry
                mbx_entry := os_mbx_get_mbx_entry (current, mbx_index);
 
-               if iterator = 0 then
+               if iterator = 1 then
                   --  if this was the first mbx, we just increase the mbx head
                   os_mbx_inc_mbx_head (current);
                else
                   --  in other case, for now we "compact" the rest of the mbx
                   --  queue, so that there is no "hole" in it for the next mbx
                   --  search.
-                  for iterator2 in (iterator + 1) ..
-                                   (os_mbx_get_mbx_count (current) - 1)
+                  for iterator2 in iterator .. os_mbx_get_mbx_count (current)
```

Comment:
I think we still need the iterator + 1

Reply:
Well I hesitated, but prefer os_mbx_index_t'Succ (iterator) rather an add operation which needs checks.

Reply:
I am OK with the Succ. It is just that we need to start iterator2 at iterator + 1 and not iterator.

Reply:
OK